src/prommis/uky/costing/ree_plant_capcost.py

```diff
@@ -2669,3 +2869,564 @@ def costing_upper_bound_eq(c, p):
+            operating_inflation_percentage: inflation rate for operating costs
+                during the operating period. Set to 0 to indicate no inflation.
+            revenue_inflation_percentage: inflation rate for revenue during the
+                operating period. Set to 0 to indicate no inflation.
+            royalty_charge_percentage_of_revenue: percentage of revenue charged
+                as royalties; ignored if royalty_expression is not None. Set to zero
+                to indicate no royalties are charged.
+            royalty_expression: set the value or expression to calculate royalties.
+        """
+
+        # input verification
+
+        input_list = [
+            b.config.total_capital_cost,
+            b.config.annual_operating_cost,
+            b.config.annual_revenue,
+            b.config.cost_year,
+        ]
+
+        if True in [
+            i is not None for i in input_list
+        ]:  # if one fixed input is passed, require all fixed inputs
+            b.verify_calculate_from_inputs()
+        elif not (
+            fixed_OM and variable_OM
+        ):  # if OM variables don't exist, fail and suggest ways to fix error
+            raise AttributeError(
+                "If capital, fixed O&M, or variable O&M costs are not calculated, "
+                "then inputs for total_capital_cost, annual_operating_cost, and annual_revenue "
+                "must be passed, and cost_year must be passed as a string, e.g. '2021'."
+                "Alternatively, set fixed_OM and variable_OM to True to calculate O&M results."
+            )
+        else:  # otherwise use expected results from costing block
+            b.verify_calculate_from_costing_block()
+
+        # check required arguments
+        QGESSCostingData.assert_config_argument_set(b, name="discount_percentage")
+        QGESSCostingData.assert_config_argument_set(b, name="plant_lifetime")
+
+        # check capital expenditure arguments
+        if (
+            b.config.has_capital_expenditure_period
+            and b.config.capital_expenditure_percentages is None
+        ):
+            b.config.capital_expenditure_percentages = [10, 60, 30]
+        elif not b.config.has_capital_expenditure_period:
+            b.config.capital_expenditure_percentages = []
+
+        if b.config.has_capital_expenditure_period:
+            QGESSCostingData.verify_percentages_list(
+                b.config, name="capital_expenditure_percentages"
+            )
+
+        # check optional expressions
+        QGESSCostingData.assert_Pyomo_object(b.config, name="royalty_expression")
+
+        # build variables
+
+        b.pv_capital_cost = Var(
+            initialize=-b.CAPEX,
+            bounds=(-1e4, 0),
+            doc="present value of total lifetime capital costs in $MM; negative cash flow",
+            units=b.cost_units,
+        )
+
+        b.pv_operating_cost = Var(
+            initialize=-b.CAPEX,
```

Comment:
Should this be initialized as -b.OPEX? (similar question for initialization of other variables, not sure if it is important here)

Reply:
CAPEX is used to initialize the PV values as a rough 1-2 order of magnitude initial guess based on user inputs. I agree that we can update the initial guesses, as there may be cases where CAPEX, OPEX and REVENUE do not have similar values.